src/lib.rs

```diff
@@ -363,253 +527,465 @@ struct Category<BlockNumber, Moment, AccountId> {
+                    category.archived = archived;
+                    archived_change_to = Some(archived);
+                }
+
+                if deleted != category.deleted {
+                    category.deleted = deleted;
+                    deleted_changed_to = Some(deleted)
+                }
+
+                // Write back mutated category
+                <CategoryById<T>>::insert(category_id, category);
+
+                // Generate event
+                Self::deposit_event(RawEvent::CategoryUpdated(category_id, archived_change_to, deleted_changed_to));
 
-        // The signature could also look like: `fn on_initialize()`
-        fn on_initialize(_n: T::BlockNumber) {
-            // Anything that needs to be done at the start of the block.
-            // We don't do anything here.
+                Ok(())
+
+            } else {
+
+                // If path is empty, it just means the category did not exist,
+                // which means the user provided an invalid category id.
+                Err(ERROR_CATEGORY_DOES_NOT_EXIST)
+            }
         }
 
-        // The signature could also look like: `fn on_finalize()`
-        fn on_finalize(_n: T::BlockNumber) {
-            // Anything that needs to be done at the end of the block.
-            // We just kill our dummy storage item.
-            <Dummy<T>>::kill();
+        /// Create new thread in category
+        fn create_thread(origin, category_id: CategoryId, title: Vec<u8>, text: Vec<u8>) -> dispatch::Result {
+
+            /*
+             * Update SPEC with new errors,
+             * and mutation of Category class,
+             * as well as side effect to update Category::num_threads_ever_created.
+             */ 
+
+            // Check that its a valid signature
+            let who = ensure_signed(origin)?;
+
+            Self::ensure_is_forum_member(who)?;
+
+            // Get path from parent to root of category tree.
+            let mut category_tree_path = vec![];
+
+            Self::build_category_tree_path(category_id, &mut category_tree_path);
+
+            // ....
+            Self::ensure_can_mutate_in_category_path_tip(&category_tree_path)?;
+            
+            // Validate title
+            ensure_thread_title_is_valid(&title)?;
+
+            // Validate post text
+            ensure_post_text_is_valid(&text)?;
+
+            /*
+             * Here it is safe to mutate state.
+             */
+
+            // Get category
+            //let category = category_tree_path.first()
+
+            // threadById extended with new Thread instance under old value of nextThreadId as identifier
+
+            let next_thread_id = <NextThreadId<T>>::get();
+
+            let new_thread = Thread {
+                id : next_thread_id,
+                title : title.clone(),
+                category_id: category_id,
+                thread_nr: 777777,
+                moderation : None,
+                num_unmoderated_posts: 0,
```

Comment:
Seems like a typo:
in this case, `num_unmoderated_posts` should be `1`
and `num_moderated_posts` should be `0`.

Reply:
Discarded, premature review.